src/cmd.rs

```diff
@@ -70,30 +71,30 @@ impl<'a, T: FromRedisValue> Iterator for Iter<'a, T> {
 #[cfg(feature = "aio")]
 use crate::aio::ConnectionLike as AsyncConnection;
 
-/// Represents a redis iterator that can be used with async connections.
+/// The inner future of AsyncIter
 #[cfg(feature = "aio")]
-pub struct AsyncIter<'a, T: FromRedisValue + 'a> {
+struct AsyncIterInner<'a, T: FromRedisValue + 'a> {
     batch: std::vec::IntoIter<T>,
     con: &'a mut (dyn AsyncConnection + Send + 'a),
     cmd: Cmd,
 }
 
+/// Represents the state of AsyncIter
 #[cfg(feature = "aio")]
-impl<'a, T: FromRedisValue + 'a> AsyncIter<'a, T> {
-    /// ```rust,no_run
-    /// # use redis::AsyncCommands;
-    /// # async fn scan_set() -> redis::RedisResult<()> {
-    /// # let client = redis::Client::open("redis://127.0.0.1/")?;
-    /// # let mut con = client.get_async_connection().await?;
-    /// con.sadd("my_set", 42i32).await?;
-    /// con.sadd("my_set", 43i32).await?;
-    /// let mut iter: redis::AsyncIter<i32> = con.sscan("my_set").await?;
-    /// while let Some(element) = iter.next_item().await {
-    ///     assert!(element == 42 || element == 43);
-    /// }
-    /// # Ok(())
-    /// # }
-    /// ```
+enum IterOrFuture<'a, T: FromRedisValue + 'a> {
```

Comment:
I feel like this introduces quite a bit of complexity that doesn't seem necessary from first principles. It seems to me that we should be able to write an `AsyncIter` (I guess with a hand-written `Future` impl) that doesn't need the intermediate `AsyncIterInner` or `IterOrFuture` stuff, but mainly has the `AsyncConnection` as internal state.

Is there some reason for not going that route that I'm missing? Is there a good reason to batch stuff that I'm not seeing? (Other than, what the code currently looks like.)

Reply:
I don't know why this was implemented using batching, I guess that is following the pattern of sync iterator.

From what I understanding, `ConnectionLike::req_packed_command` returns in batch already.

Then, with a hand crafted future, wouldn't be the same issue? since we need to keep the future while we poll it, until it returns `Poll::ready`, but maybe I'm missing something here since it's my first time using future's api.

Reply:
Yes, that's what I mean, shouldn't be a problem internally, but since it's a public trait, then will break compatibility with other crates/binaries that implement `ConnectionLike`, right? and I don't know if there's any policy on when to break compat in redis-rs, since still doesn't hit 1.0 release, it means that we can't be explicit in the version for the breakage

Reply:
RE batching: I'm probably missing something in the discussion here, but is there any way to avoid batching in the implementation, at least without something significantly more complicated? The current (albeit broken) impl is simply following the procedure for using these commands, e.g., `SCAN`, which returns results in batches and requires you to specify the next "cursor" value in order to get the next batch. 

I'm not sure how we would intelligently pull in the batches from the server and stream them out except by doing something approximately like what's been implemented in this PR.

Sorry for being extremely late to the party here, but as the problem has come up again of late, I think we should consider reviving these changes and getting them in. I'm also not clear why there are discussions of making breaking changes or how bypassing `req_packed_command` would accomplish anything.

Reply:
I also couldn't understand why bypassing `req_packed_command` was necessary. Since the server returns results in batches, and we want an interface that yields one result at a time, we need some kind of flattening, the `Iter` variant (the buffer) is definitely required.

Or there's another solution where we stream batches of results (Item=Vec<T>), and ask the user to flatten the stream manually. This approach may eliminate the intermediate state completely, but is apparently a breaking change.

Personally I would vote up the current implementation of this PR.